pkg/broker/broker.go

```diff
@@ -850,10 +859,194 @@ func (a AnsibleBroker) Unbind(
+	//   -> PlanID            uuid.UUID
+	//   -> ServiceID         uuid.UUID
+	// ServiceID maps directly to a Spec.Id found in etcd. Can pull Spec via
+	// Dao::GetSpec(id string)
+
+	//-> Parameters        map[string]string
+	// User provided configuration answers for the AnsibleApp
+
+	// -> AcceptsIncomplete bool
+	// true indicates both the SC and the requesting client (sc client). If param
+	// is not included in the req, and the broker can only provision an instance of
+	// the request plan asyncronously, broker should reject with a 422
+	// NOTE: Spec.Async should indicate what level of async support is available for
+	// a given ansible app
+
+	//}
+
+	// Summary:
+	// For our purposes right now, the ServiceID and the Params should be enough to
+	// Update an ansible app.
+	////////////////////////////////////////////////////////////
+	// Update Flow
+	// -> Retrieve Spec from etcd (if missing, 400, this returns err missing)
+	// -> Retrieve Instance from etcd (if missing, 400, this returns err missing)
+	// -> TODO: Check to see if the spec supports or requires async, and reconcile
+	//    need a typed error condition so the REST server knows correct response
+	//    depending on the scenario
+	//    (async requested, unsupported, 422)
+	//    (async not requested, required, ?)
+	// -> Update entry in /instance, ID'd by instance. Value should be Instance type
+	//    Purpose is to make sure everything neeed to deprovision is available
+	//    in persistence.
+	// -> Update!
+	////////////////////////////////////////////////////////////
+
+	/*
+	   dao GET returns error strings like CODE: message (entity) [#]
+	   dao SetServiceInstance returns what error?
+	   dao.SetState returns what error?
+	   Provision returns what error?
+	   SetExtractedCredentials returns what error?
+
+	   broker
+	   * normal synchronous return UpdateResponse
+	   * normal async return UpdateResponse
+
+	   handler returns the following
+	   * synchronous update return 201 created
+	   * instance already exists with IDENTICAL parameters to existing instance, 200 OK
+	   * async provision 202 Accepted
+	   * if only support async and no accepts_incomplete=true passed in, 422 Unprocessable entity
+
+	*/
+
+	var err error
+
+	si, err := a.dao.GetServiceInstance(instanceUUID.String())
+	if err != nil {
+		a.log.Debug("Error retrieving instance")
+		return nil, ErrorNotFound
+	}
+
+	// If no plan was specified we'll continue with the current plan
+	if req.PlanID == "" {
+		req.PlanID = (*si.Parameters)["_apb_plan_id"].(string)
+	}
+
+	// Retrieve requested spec
+	spec, err := a.dao.GetSpec(si.Spec.ID)
+	if err != nil {
+		// etcd return not found i.e. code 100
+		if client.IsKeyNotFound(err) {
+			return nil, ErrorNotFound
+		}
+		// otherwise unknown error bubble it up
+		return nil, err
+	}
+
+	// Make sure that we can find the plan we're updating from. This should probably never fail, but cover our tail.
```

Comment:
I would abstract the ladder of validations that occurs here into their own, testable functions. Makes things read on a nice high level for the uninitiated and hides the ugly details, while also letting us easily change or add to their implementation without changing the core flow that will likely remain the same:

```
if err := validateCurrentPlan(); err != nil {
		a.log.Error("The plan %s, specified for updating from on instance %s, does not exist.", (*si.Parameters)["_apb_plan_id"], si.ID)
 		return nil, ErrorPlanNotFound
}

if err := validateNextPlan(); err != nil {
		a.log.Error("The plan %s, specified for updating from on instance %s, does not exist.", (*si.Parameters)["_apb_plan_id"], si.ID)
 		return nil, ErrorPlanNotFound
}

// Make sure that the current plan supports updating to the requested plan if a plan update was requested.
if err := validatePlanTransition(currentPlan, nextPlan); err != nil {
// Invalid Transition! throw error.
}

// Make sure that parameters requested for update exist on the new plan and that they are updatable.
if err := validateNextParameters(param, plan); err != nil { /* handle */ }

// elsewhere
func validateCurrentPlan(...) {
// Iterate spec plans and make sure the plan pulled from the si params actually exists
}

func validateNextPlan() {
// Iterate and confirm
		a.log.Error("The plan %s, specified for updating to on instance %s, does not exist.", req.PlanID, si.ID)
		return nil, ErrorPlanNotFound
}

// Etc...
```